File: January2021/concatenation-of-consecutive-binary-numbers.java
class Solution {
    public int concatenatedBinary(int n) {
        int mod = 1000000007 ;
        int newNo =0;

        for(int num=1;num<=n;num++){

            String binaryRep= Integer.toBinaryString(num);

            for(char ch : binaryRep.toCharArray()){
                int val= (ch=='0')?0:1;

                newNo= ((newNo*2)%mod + val) % mod;
            }
        }

        return newNo;
    }
}

File: September2022/C++/ConcatenationOfConsecutiveBinaryNumbers.cpp
class Solution {
private:
    int getBinaryNumberLength(int n) {
         int length = 0;
         while (n > 0) {
             n /= 2;
             length++;
         }
         return length;
    }
    
public:
    // TC: O(nlogn)
    // SC: O(1)
    int concatenatedBinary(int n) {
        int MOD = (int) 1e9 + 7;

        long ans = 0;

        for (int i = 1; i <= n; i++) {
            int noOfShifts = getBinaryNumberLength(i);
            ans = (ans << noOfShifts) % MOD;
            ans += i;
            ans %= MOD;
        }

        return (int) ans % MOD;
    }
};
